Netcdf export

gloriajjl opened the pull request:
Implementation of the NetCDF export issue #71

gloriajjl committed:
NetCDF export #71

gloriajjl committed:
Merge branch 'dev' of https://github.com/IonosatMicro/promis into dev

gloriajjl committed:
Added netcdf4 and numpy libraries to the requirements

gloriajjl commented on backend/promis/export.py:
```diff
@@ -118,6 +116,72 @@ def csv_export(table, datalabel="Data", dataunits="units"):
+
+   from netCDF4 import Dataset
+
+   dataset = Dataset("temp.nc", 'w', format='NETCDF4')
```
Я не впевнена щодо того, чи треба тут створювати файл з унікальним ім'ям чи ні. Поки що це один файл. І я його не видаляю по закінченню роботи. Все одно перезапишеться наступного разу.

landswellsong replied:
Проблема тут в тому, що на релізній версії буде `gunicorn` і він запускатиме процеси паралельно. Коли у тебе два потоки будуть користувати той сами файл, буде не альо. Раджу зробити [через отак](https://docs.python.org/3.7/library/tempfile.html) щоб файл був гарантовано унікальний, якщо уже не можна уникнути файла. А без файла там ніяк?

gloriajjl replied:
Без файла не получилося :( Я пробувала через StringIO зробити, але потім почитала, що здається такий фокус не прокатує з  Dataset

landswellsong replied:
Я пропоную тоді так: зробити зараз з файлом і зробити міні-ішью ще на «зробити без файла» з тегом `improvement`, на колись потім.

landswellsong commented on frontend/app/components/SearchResults.js:
```diff
@@ -58,7 +58,8 @@ class DataSection extends Component {
             a.download = this.state.main + '.txt';
-            a.href = '/en/api/download/' + this.props.mid + '/data/?format=txt' + src + time;
+            //a.href = '/en/api/download/' + this.props.mid + '/data/?format=txt' + src + time;
+            a.href = '/en/api/download/' + this.props.mid + '/data/?format=netcdf' + src + time;
```
Нам давно пора зробить вибір формату :) Але на це є іш'ю вже.

landswellsong commented on backend/promis/export.py:
```diff
@@ -118,6 +116,72 @@ def csv_export(table, datalabel="Data", dataunits="units"):
+   altitudeVar.units = 'km'
+   dataVar.units = 'Date (YYYYDDD), "UT (ms), Longitude (deg), Latitude (deg), Altitude (km)'
+
+   dateVar[:] = dates
```
Дивний синтаксис, це для примусового копіювання?

gloriajjl commented on backend/promis/export.py:
```diff
@@ -118,6 +116,72 @@ def csv_export(table, datalabel="Data", dataunits="units"):
+   latitudeVar.units = 'deg'
+   longtitudeVar.units = 'deg'
+   altitudeVar.units = 'km'
+   dataVar.units = 'Date (YYYYDDD), "UT (ms), Longitude (deg), Latitude (deg), Altitude (km)'
```
Мене насправді непокоїть ще оцей рядок. Я прочитала десь, але тепер не можу знайти де, що ніби для compoundType не можна задавати атрибути, в тому числі units для складових цього великого основного типу. Тому я зробила так. Але не впевнена в своїй правоті

landswellsong replied:
Це в NetCDF власне? Зараз сходу не можу сказати, попробую зібрати і перевірити.

gloriajjl commented on backend/promis/export.py:
```diff
@@ -118,6 +116,72 @@ def csv_export(table, datalabel="Data", dataunits="units"):
+   data = dataset.createDimension('Data', numElems)
+
+   import numpy as np
+   dateVar= dataset.createVariable('Date', np.float32, ('Date'))
```
Не зовсім впевнена щодо типу даних Дата. Тут я поставила float32 зразу, і забула виправити перед комітом. Мабуть логічніше було int32. Алн можливо це має бути щось зовсім інше? String чи взагалі якийсь тип Дата чи Час. Далі в compoundType (рядок 152) я ці дані записую як стрінг. Але мабуть треба зробити тип однаковим тут і там. питання тільки в тому - яким він має бути. В netcdf4 бібліотеці є якісь інструменти для конвертування дати та часу в гарне представлення, але я ними не користувалася, бо не знала, чи потрібно це

landswellsong replied:
Можна пушати в ту саму гілку і коміт додастся :)

gloriajjl committed:
- changed 'date' type to int32 for now
- changed hardcoded filename to automatic generation of the temp files with random names

gloriajjl committed:
Changed data representation in the netcdf file

gloriajjl committed:
got rid of some extra copying operations

gloriajjl committed:
Measuring time is seconds for the data with sampling frequency 1Hz, milliseconds - for 1kHz, and microseconds for more

gloriajjl committed:
Got rid of Server error for netcdf format

elpiankova reviewed (changes requested):
Unique name of temporary netcdf file and add suport for multicomponent data like for Variant

gloriajjl committed:
Merge branch 'dev' into netcdf_export

gloriajjl committed:
Added support for the multicomponenet data
